### PR TITLE
Join Section | Backend API

### PR DESCRIPTION
**Pull request by RitikJaiswal75**
## Description
This PR configures routes for uploading the fetching the data for Join Section.

### Flow
- As the user clicks on submit button on the Front end Join Section after finishing the form. 
- The data hits the backend at `/user/:userId/intro` as a post request.
- This data is then sent to firebase, which is uploaded to the firestore database.
- Once the data is stored the User gets a link to his/her information that is stored which can then be shared with the superUser (Ankush) 
- On hitting the same link with a get request as SuperUser the data is returned from firestore.

### What did I do?
- Configured a route `/users/:userId/intro` with `GET` to get user data by superuser.
- Configured a route `/users/self/intro` with `POST` to send the user data that is to be saved in the database.
- Created a model for uploading and fetching the data.

**This PR closes issue #714** 

### References for the Join Section details and discussions:
- https://github.com/Real-Dev-Squad/website-www/issues/280
- https://github.com/Real-Dev-Squad/website-www/issues/286

### CODE REVIEW

File: controllers/users.js [hunk 1]
```diff
@@ -292,6 +292,19 @@ const rejectProfileDiff = async (req, res) => {
   }
 };
 
+const addUserIntro = async (req, res) => {
+  await userQuery.addJoinData(req.body);
+  return res.json({ message: "User Data Added Succesfully" });
```
Comment (vinayak-trivedi) [inline controllers/users.js:297]: please put this inside a try catch block, and handle the errors correctly

File: controllers/users.js [hunk 2]
```diff
@@ -292,6 +292,19 @@ const rejectProfileDiff = async (req, res) => {
...
+};
+
+const getUserIntro = async (req, res) => {
+  const data = await userQuery.getJoinData(req.params.username);
+  return res.json({
+    message: "Got user data",
+    data: data,
+  });
+};
```
Comment (vinayak-trivedi) [inline controllers/users.js:306]: put this also in a try catch block, and the message can be changed to something like `user data returned`

File: models/users.js [hunk 3]
```diff
@@ -6,6 +6,8 @@ const walletConstants = require("../constants/wallets");
 
 const firestore = require("../utils/firestore");
 const { fetchWallet, createWallet } = require("../models/wallets");
+const db = require("../utils/firestore");
```
Comment (vinayak-trivedi) [inline models/users.js:9]: why are you importing it here, it's already imported above in line 7

File: models/users.js [hunk 4]
```diff
@@ -56,6 +58,25 @@ const addOrUpdate = async (userData, userId = null) => {
   }
 };
 
+const addJoinData = async (userData) => {
+  try {
+    await db.collection("joining_data").doc(userData.userName).set(userData);
```
Comment (vinayak-trivedi) [inline models/users.js:63]: instead of doing this you can have a variable which contains the value `firestore.collection("joining-data")` and use it here directly that would be much cleaner, one more thing, can we change the name to `intro` from joining data that is better in my opinion

**Review by vinayak-trivedi — CHANGES_REQUESTED** [review]

**Comment by vinayak-trivedi** [comment]
one more thing please write test for it as well @RitikJaiswal75

File: models/users.js [hunk 5]
```diff
@@ -56,6 +58,25 @@ const addOrUpdate = async (userData, userId = null) => {
   }
 };
 
+const addJoinData = async (userData, username) => {
+  try {
+    await joinModel.doc(username).set(userData);
```
Comment (vinayak-trivedi) [inline models/users.js:63]: Hey ritik don't use username for doc id if the user changes his username, things might break

**Review by vinayak-trivedi — CHANGES_REQUESTED** [review]

File: controllers/users.js [hunk 6]
```diff
@@ -292,6 +292,29 @@ const rejectProfileDiff = async (req, res) => {
   }
 };
 
+const addUserIntro = async (req, res) => {
+  try {
+    await userQuery.addJoinData(req.body);
```
Comment (prakashchoudhary07) [inline controllers/users.js:297]: Is the body being sanitized here?

File: models/users.js [hunk 7]
```diff
@@ -6,7 +6,9 @@ const walletConstants = require("../constants/wallets");
 
 const firestore = require("../utils/firestore");
 const { fetchWallet, createWallet } = require("../models/wallets");
+const logger = require("../utils/logger");
```
Comment (prakashchoudhary07) [inline models/users.js:9]: Do we need to import this?

File: controllers/users.js [hunk 8]
```diff
@@ -292,6 +292,29 @@ const rejectProfileDiff = async (req, res) => {
...
+  } catch (err) {
+    logger.error("Could not save user data");
+    throw err;
+  }
+};
+
+const getUserIntro = async (req, res) => {
+  try {
+    const data = await userQuery.getJoinData(req.params.userId);
+    return res.json({
```
Comment (prakashchoudhary07) [inline controllers/users.js:308]: What happens when userId in not valid?

File: routes/users.js [hunk 9]
```diff
@@ -15,7 +15,10 @@ router.get("/self", authenticate, users.getSelfDetails);
 router.get("/isUsernameAvailable/:username", authenticate, users.getUsernameAvailabilty);
 router.get("/chaincode", authenticate, users.generateChaincode);
 router.get("/:username", users.getUser);
+router.get("/:userId/intro", users.getUserIntro);
 // upload.single('profile') -> multer inmemory storage of file for type multipart/form-data
+
+router.post("/:userId/intro", users.addUserIntro);
```
Comment (prakashchoudhary07) [inline routes/users.js:21]: Is that comment related to your changes?

File: models/users.js [hunk 10]
```diff
@@ -56,6 +58,32 @@ const addOrUpdate = async (userData, userId = null) => {
...
+const addJoinData = async (userData) => {
+  try {
+    await joinModel.add(userData);
+  } catch (err) {
+    logger.error("Error in adding data", err);
+    throw err;
+  }
+};
+
+const getJoinData = async (userid) => {
```
Comment (prakashchoudhary07) [inline models/users.js:70]: Naming conventions for userid
`userid` -> `userId`

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: routes/users.js (hunk 9 above)
Comment (RitikJaiswal75) [reply routes/users.js:21]: No, that comment exists in code i did not add or delete it

**Review by Maheima — DISMISSED** [review]
LGTM!

**Review by rupal-2505 — DISMISSED** [review]
LGTM!

File: controllers/users.js [hunk 11]
```diff
@@ -292,6 +292,64 @@ const rejectProfileDiff = async (req, res) => {
   }
 };
 
+const addUserIntro = async (req, res) => {
+  try {
+    if (req.params.userId === req.body.userId) {
```
Comment (vinayak-trivedi) [inline controllers/users.js:297]: I didn't get why we need to check this here

Comment (vinayak-trivedi) [reply controllers/users.js:297]: okay got it

**Review by vinayak-trivedi — APPROVED** [review]
LGTM 👍

File: models/users.js [hunk 12]
```diff
@@ -56,6 +57,32 @@ const addOrUpdate = async (userData, userId = null) => {
...
+  } catch (err) {
+    logger.error("Error in adding data", err);
+    throw err;
+  }
+};
+
+const getJoinData = async (userId) => {
+  try {
+    const userData = [];
+    const joinData = await joinModel.where("userId", "==", userId).limit(1).get();
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply models/users.js:72]: possibly we may get more than one filled by the same person. So we may need this in the feature.

File: controllers/users.js [hunk 13]
```diff
@@ -292,6 +292,64 @@ const rejectProfileDiff = async (req, res) => {
...
+      await userQuery.addJoinData(data);
+
+      return res.json({ message: "User Data Added Succesfully" });
+    } else {
+      return res.status(401).json({ message: "Unauthorized request" });
+    }
+  } catch (err) {
+    logger.error("Could not save user data");
+    throw err;
+  }
```
Comment (prakashchoudhary07) [inline controllers/users.js:331]: Shouldn't you have returned error with 5XX response code?

Comment (prakashchoudhary07) [inline controllers/users.js:331]: Shouldn't this return 5XX errror ?

File: controllers/users.js [hunk 14]
```diff
@@ -292,6 +292,64 @@ const rejectProfileDiff = async (req, res) => {
...
+      });
+    } else {
+      return res.status(404).json({
+        message: "Data Not Found",
+      });
+    }
+  } catch (err) {
+    logger.error("Could Not Get User Data", err);
+    throw err;
+  }
```
Comment (prakashchoudhary07) [inline controllers/users.js:350]: Shouldn't this return a 5XX error?

File: routes/users.js [hunk 15]
```diff
@@ -15,6 +16,9 @@ router.get("/self", authenticate, users.getSelfDetails);
 router.get("/isUsernameAvailable/:username", authenticate, users.getUsernameAvailabilty);
 router.get("/chaincode", authenticate, users.generateChaincode);
 router.get("/:username", users.getUser);
+router.get("/:userId/intro", authenticate, authorizeRoles([SUPERUSER]), users.getUserIntro);
+router.post("/:userId/intro", authenticate, validateJoinData, users.addUserIntro);
+
```
Comment (prakashchoudhary07) [inline routes/users.js:20]: Will userId change, if so then can't any user can add data by changing userId?

File: test/integration/users.test.js [hunk 16]
```diff
@@ -14,6 +14,8 @@ const profileDiffData = require("../fixtures/profileDiffs/profileDiffs")();
 const superUser = userData[4];
 
 const config = require("config");
+const join = require("../fixtures/user/join");
+const { addJoinData } = require("../../models/users");
```
Comment (prakashchoudhary07) [inline test/integration/users.test.js:18]: What is join here?

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: controllers/users.js (hunk 11 above)
Comment (prakashchoudhary07) [inline controllers/users.js:297]: Can't any user add data here, if userId in params and body is matched?

**Review by rupal-2505 — DISMISSED** [review]

**Review by Maheima — APPROVED** [review]
LGTM!

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: routes/users.js [hunk 17]
```diff
@@ -15,6 +15,9 @@ router.get("/self", authenticate, users.getSelfDetails);
 router.get("/isUsernameAvailable/:username", authenticate, users.getUsernameAvailabilty);
 router.get("/chaincode", authenticate, users.generateChaincode);
 router.get("/:username", users.getUser);
+router.get("/:userId/intro", authenticate, authorizeRoles([SUPERUSER]), users.getUserIntro);
+router.post("/:userId/intro", authenticate, userValidator.validateJoinData, users.addUserIntro);
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply routes/users.js:19]: i need it for get as well, as userid is used for get request

[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply routes/users.js:19]: The super user sends a get request, If I do not pass the user id of a specific user, how will the super user get that user's data?

I can agree to remove the userId parameter from the post request but not from the get request.

[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply routes/users.js:19]: I am talking about the post request

line 18 got selected by mistake

File: controllers/users.js [hunk 18]
```diff
@@ -292,6 +292,64 @@ const rejectProfileDiff = async (req, res) => {
   }
 };
 
+const addUserIntro = async (req, res) => {
+  try {
+    if (req.userData.id === req.body.userId) {
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply controllers/users.js]: Resolved as per the new push

File: test/integration/users.test.js [hunk 19]
```diff
@@ -300,6 +302,108 @@ describe("Users", function () {
...
+          expect(res.body.message).to.equal("User Data Added Succesfully");
+          return done();
+        });
+    });
+
+    it("Should return 401 for adding data of other user", function (done) {
+      chai
+        .request(app)
+        .post(`/users/${userId}/intro`)
+        .set("cookie", `${cookieName} = ${jwt}`)
```
[the remark this answers is not in the document]
Comment (RitikJaiswal75) [reply test/integration/users.test.js]: Resolved as per the new push.

**Review by prakashchoudhary07 — APPROVED** [review]
LGTM, thanks for making the requested changes and for the tests :rocket:

**Review by Maheima — APPROVED** [review]

**Review by vinayak-trivedi — APPROVED** [review]
LGTM 👍
